File: momentra/momentra/Shell/PersonalEmpty/PersonalExpenseCategoryCatalog.swift
import Foundation

enum PersonalExpenseCategoryCatalog {
    struct Category: Identifiable {
        let id: String
        let code: String
        let label: String
        let emoji: String
        let subcategories: [Subcategory]
    }

    struct Subcategory: Identifiable {
        let id: String
        let code: String
        let label: String
    }

    static let masterCategories: [Category] = [
        Category(id: "FOOD", code: "FOOD", label: "Food", emoji: "🍕", subcategories: [
            Subcategory(id: "FOOD_DINING", code: "FOOD_DINING", label: "Food & Dining"),
            Subcategory(id: "CAFE", code: "CAFE", label: "Cafe"),
            Subcategory(id: "GROCERIES", code: "GROCERIES", label: "Groceries"),
            Subcategory(id: "RESTAURANT", code: "RESTAURANT", label: "Restaurant"),
        ]),
        Category(id: "TRANSPORT", code: "TRANSPORT", label: "Transport", emoji: "🚗", subcategories: [
            Subcategory(id: "TRANSPORT", code: "TRANSPORT", label: "Transport"),
            Subcategory(id: "FUEL", code: "FUEL", label: "Fuel"),
            Subcategory(id: "RIDE", code: "RIDE", label: "Ride share"),
        ]),
        Category(id: "SHOPPING", code: "SHOPPING", label: "Shopping", emoji: "🛍️", subcategories: [
            Subcategory(id: "SHOPPING", code: "SHOPPING", label: "Shopping"),
            Subcategory(id: "CLOTHING", code: "CLOTHING", label: "Clothing"),
            Subcategory(id: "ELECTRONICS", code: "ELECTRONICS", label: "Electronics"),
        ]),
        Category(id: "CAFE", code: "CAFE", label: "Cafe", emoji: "☕", subcategories: [
            Subcategory(id: "CAFE", code: "CAFE", label: "Cafe"),
            Subcategory(id: "COFFEE", code: "COFFEE", label: "Coffee"),
        ]),
        Category(id: "HEALTH", code: "HEALTH", label: "Health", emoji: "💊", subcategories: [
            Subcategory(id: "HEALTH", code: "HEALTH", label: "Health"),
            Subcategory(id: "PHARMACY", code: "PHARMACY", label: "Pharmacy"),
            Subcategory(id: "FITNESS", code: "FITNESS", label: "Fitness"),
        ]),
        Category(id: "ENTERTAINMENT", code: "ENTERTAINMENT", label: "Entertainment", emoji: "🎬", subcategories: [
            Subcategory(id: "ENTERTAINMENT", code: "ENTERTAINMENT", label: "Entertainment"),
            Subcategory(id: "STREAMING", code: "STREAMING", label: "Streaming"),
            Subcategory(id: "EVENTS", code: "EVENTS", label: "Events"),
        ]),
        Category(id: "BILLS", code: "BILLS", label: "Bills", emoji: "🧾", subcategories: [
            Subcategory(id: "BILLS", code: "BILLS", label: "Bills"),
            Subcategory(id: "UTILITIES", code: "UTILITIES", label: "Utilities"),
            Subcategory(id: "RENT", code: "RENT", label: "Rent"),
            Subcategory(id: "HOUSING", code: "HOUSING", label: "Housing"),
        ]),
        Category(id: "OTHER", code: "OTHER", label: "Other", emoji: "📦", subcategories: [
            Subcategory(id: "OTHER", code: "OTHER", label: "Other"),
            Subcategory(id: "MISC", code: "MISC", label: "Miscellaneous"),
        ]),
    ]

    static func labelForCode(_ code: String?) -> String {
        guard let code, !code.isEmpty else { return "Other" }
        let upper = code.uppercased()
        for cat in masterCategories {
            if let sub = cat.subcategories.first(where: { $0.code == upper }) { return sub.label }
            if cat.code == upper { return cat.label }
        }
        return code.replacingOccurrences(of: "_", with: " ").capitalized
    }

    static func encodeCategoryCode(categoryCode: String, subcategoryCode: String?) -> String {
        guard let sub = subcategoryCode, !sub.isEmpty else { return categoryCode.uppercased() }
        return sub.uppercased()
    }

    static func decodeFromStored(_ stored: String?) -> (String, String?) {
        guard let stored, !stored.isEmpty else { return ("OTHER", nil) }
        let upper = stored.uppercased()
        for cat in masterCategories {
            if let sub = cat.subcategories.first(where: { $0.code == upper }) {
                return (cat.code, sub.code)
            }
        }
        return (upper, nil)
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/personal/shared/PersonalExpenseCategoryCatalog.kt
package com.example.momentra.ui.shell.personal.shared

/** Static category catalog for Master Expense + Edit Transaction (no backend catalog yet). */
object PersonalExpenseCategoryCatalog {
    data class Category(
        val code: String,
        val label: String,
        val emoji: String,
        val subcategories: List<Subcategory>,
    )

    data class Subcategory(
        val code: String,
        val label: String,
    )

    val masterCategories: List<Category> = listOf(
        Category("FOOD", "Food", "🍕", listOf(
            Subcategory("FOOD_DINING", "Food & Dining"),
            Subcategory("CAFE", "Cafe"),
            Subcategory("GROCERIES", "Groceries"),
            Subcategory("RESTAURANT", "Restaurant"),
        )),
        Category("TRANSPORT", "Transport", "🚗", listOf(
            Subcategory("TRANSPORT", "Transport"),
            Subcategory("FUEL", "Fuel"),
            Subcategory("RIDE", "Ride share"),
        )),
        Category("SHOPPING", "Shopping", "🛍️", listOf(
            Subcategory("SHOPPING", "Shopping"),
            Subcategory("CLOTHING", "Clothing"),
            Subcategory("ELECTRONICS", "Electronics"),
        )),
        Category("CAFE", "Cafe", "☕", listOf(
            Subcategory("CAFE", "Cafe"),
            Subcategory("COFFEE", "Coffee"),
        )),
        Category("HEALTH", "Health", "💊", listOf(
            Subcategory("HEALTH", "Health"),
            Subcategory("PHARMACY", "Pharmacy"),
            Subcategory("FITNESS", "Fitness"),
        )),
        Category("ENTERTAINMENT", "Entertainment", "🎬", listOf(
            Subcategory("ENTERTAINMENT", "Entertainment"),
            Subcategory("STREAMING", "Streaming"),
            Subcategory("EVENTS", "Events"),
        )),
        Category("BILLS", "Bills", "🧾", listOf(
            Subcategory("BILLS", "Bills"),
            Subcategory("UTILITIES", "Utilities"),
            Subcategory("RENT", "Rent"),
            Subcategory("HOUSING", "Housing"),
        )),
        Category("OTHER", "Other", "📦", listOf(
            Subcategory("OTHER", "Other"),
            Subcategory("MISC", "Miscellaneous"),
        )),
    )

    fun categoryForCode(code: String?): Category? {
        val normalized = code?.uppercase()?.substringBefore('_') ?: return null
        return masterCategories.find {
            it.code == normalized || it.subcategories.any { s -> s.code == code?.uppercase() }
        }
    }

    fun labelForCode(code: String?): String {
        if (code.isNullOrBlank()) return "Other"
        val upper = code.uppercase()
        masterCategories.forEach { cat ->
            cat.subcategories.find { it.code == upper }?.let { return it.label }
            if (cat.code == upper) return cat.label
        }
        return code.replace('_', ' ').lowercase().replaceFirstChar { it.titlecase() }
    }

    fun subcategoryLabel(code: String?): String? {
        if (code.isNullOrBlank()) return null
        val upper = code.uppercase()
        masterCategories.forEach { cat ->
            cat.subcategories.find { it.code == upper }?.let { return it.label }
        }
        if (upper.contains('_')) {
            return upper.substringAfter('_').replace('_', ' ').lowercase()
                .replaceFirstChar { it.titlecase() }
        }
        return null
    }

    /** Encode category + subcategory for API categoryCode. */
    fun encodeCategoryCode(categoryCode: String, subcategoryCode: String?): String {
        val sub = subcategoryCode?.takeIf { it.isNotBlank() } ?: return categoryCode.uppercase()
        return sub.uppercase()
    }

    fun decodeFromStored(stored: String?): Pair<String, String?> {
        if (stored.isNullOrBlank()) return "OTHER" to null
        val upper = stored.uppercase()
        masterCategories.forEach { cat ->
            cat.subcategories.find { it.code == upper }?.let {
                return cat.code to it.code
            }
        }
        return upper to null
    }
}
